packages/core-compat-api/src/components/SystemIcon.tsx

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ComponentProps } from 'react';
+import { useApp, IconComponent } from '@backstage/core-plugin-api';
+import { compatWrapper } from '../compatWrapper';
+
+/**
+ * @public
+ * Props for the SystemIcon component.
+ */
+export type SystemIconProps = ComponentProps<IconComponent> & {
+  // The id of the system icon to render.
+  id: string;
+  // An optional fallback icon component to render when the system icon is not found.
+  // Default to () => null.
+  Fallback?: IconComponent;
+};
+
+function SystemIcon(props: SystemIconProps) {
+  const { id, Fallback = () => null, ...rest } = props;
+  const app = useApp();
+  const Component = app.getSystemIcon(id) ?? Fallback;
+  return <Component {...rest} />;
+}
+
+/**
+ * @public
+ * SystemIcon is a component that renders a system icon by its id.
+ * @example
+ * Rendering the "kind:api" icon:
+ * ```tsx
+ * <SystemIcon id="kind:api" />
```

Comment:
One thing I've wanted to do for a while with the existing icons was to allow differentiation by `spec.type`, kind=Resource and kind=Component components aren't super helpful in the catalog graph. I ended up sticking more information into the entityRef/name for visibility when a better icon granularity might have sufficed. Seems like there's a similar ask for search, #12876. Is that something we could do with this wrapper? Allowing either an entity/entityRef as input or a more specific id?

Reply:
Should the presentation API use this API for icons or are these separate concepts?

Reply:
My understanding is that for me it is a temporary compatibility component. We have a new Icons API and also a new Components API, so my initial understanding is that at some point we will deprecate this component and switch to using one of these two APIs. But let me add @Rugvip to the conversation as well to confirm 🙂

Reply:
Ah no the idea is actually for this component to stick around. It's just a lot simpler to use this than having to juggle the icons api in many places.

I wonder if what we actually need here is the ability to specify some fallbacks or to have that be some implicit behavior? For example one might specify `entity:component:service`, `entity:component`, and `entity`, and it uses the first one that is available? Or for example colons have special treatment so that specifying `entity:component:service` will look for `entity:component` and `entity` too?

Reply:
Just updated the logic, could you review again?

Reply:
Just to clarify, this component won't / shouldn't be used for the presentation API implementation? This component will be deprecated soon and I'd have to do something similar myself?

Reply:
> As a matter of fact, the kind icons map can be used as input to the default renderer, but that's a separate issue and just a historical artifact. Making a system icon renderer that then consumes through the presentation api ... sounds like a weird circular situation :)

My proposal is exactly that, moving away from the kind icons map towards using `SystemIcon` for everything. Maybe I'm just confused by the use of `kind:api` as both a system icon and an entity icon. Seems like there's overlap that exists already that we could leverage?